File: KeeperLegacy/Models/Progression.swift
import Foundation

// MARK: - Feature Flags

/// All features that can be locked/unlocked as the player progresses.
enum GameFeature: String, Codable, CaseIterable {
    case shop               = "shop"
    case habitat            = "habitat"
    case feeding            = "feeding"
    case playing            = "playing"
    case selling            = "selling"
    case customerOrders     = "customerOrders"
    case breeding           = "breeding"
    case monsterpedia       = "monsterpedia"
    case habitatExpansion   = "habitatExpansion"
    case mutations          = "mutations"
    case magicalHabitat     = "magicalHabitat"   // Story Act II gate
    case cosmetics          = "cosmetics"

    /// Minimum level required to unlock (0 = available from start)
    var requiredLevel: Int {
        switch self {
        case .shop:             return 0
        case .habitat:          return 0
        case .feeding:          return 0
        case .playing:          return 0
        case .selling:          return 1
        case .customerOrders:   return 3
        case .breeding:         return 12
        case .monsterpedia:     return 2
        case .habitatExpansion: return 2
        case .mutations:        return 15
        case .magicalHabitat:   return 0    // Level not the gate — story Act II is
        case .cosmetics:        return 5
        }
    }

    /// Story act required in addition to level (nil = no story gate)
    var requiredStoryAct: Int? {
        switch self {
        case .breeding:         return 1    // Act I must be complete
        case .magicalHabitat:   return 2
        default:                return nil
        }
    }
}

// MARK: - Milestone Rewards

struct MilestoneReward: Identifiable {
    let id: Int             // The level at which this milestone triggers
    let coinBonus: Int
    let stardustBonus: Int
    let unlocksFeature: GameFeature?
    let storyEvent: String?     // Key into Story events, if any
    let description: String

    static let milestones: [Int: MilestoneReward] = [
        5:  MilestoneReward(id: 5,  coinBonus: 200,  stardustBonus: 0,   unlocksFeature: .cosmetics,         storyEvent: nil,                   description: "Cosmetics unlocked!"),
        10: MilestoneReward(id: 10, coinBonus: 500,  stardustBonus: 25,  unlocksFeature: nil,                storyEvent: "act1_shop_secret",    description: "The shop's secret begins to reveal itself..."),
        12: MilestoneReward(id: 12, coinBonus: 300,  stardustBonus: 0,   unlocksFeature: .breeding,          storyEvent: nil,                   description: "Breeding unlocked! Requires Act I completion."),
        15: MilestoneReward(id: 15, coinBonus: 400,  stardustBonus: 0,   unlocksFeature: .mutations,         storyEvent: nil,                   description: "Mutation variants discovered!"),
        25: MilestoneReward(id: 25, coinBonus: 1000, stardustBonus: 50,  unlocksFeature: nil,                storyEvent: "act2_revelation",     description: "A revelation about the ancient civilization..."),
        50: MilestoneReward(id: 50, coinBonus: 5000, stardustBonus: 200, unlocksFeature: nil,                storyEvent: "act3_legacy",         description: "You have mastered the Keeper's Legacy!"),
    ]
}

// MARK: - Player Progression State

struct PlayerProgression: Codable {
    var currentLevel: Int
    var currentXP: Int
    var unlockedFeatures: Set<String>   // Stores GameFeature.rawValue
    var claimedMilestones: Set<Int>     // Level numbers whose rewards have been claimed
    var storyAct: Int                   // 1, 2, or 3

    init() {
        self.currentLevel       = 1
        self.currentXP          = 0
        self.unlockedFeatures   = Set(GameFeature.allCases
            .filter { $0.requiredLevel == 0 && $0.requiredStoryAct == nil }
            .map { $0.rawValue })
        self.claimedMilestones  = []
        self.storyAct           = 1
    }

    // MARK: XP / Leveling

    /// XP needed to reach the next level from the current one.
    /// Accelerated early curve, then linear from level 10+.
    var xpToNextLevel: Int {
        XPCurve.xpRequired(forLevel: currentLevel)
    }

    /// Add XP and handle level-ups. Returns an array of levels gained (usually 0 or 1).
    mutating func addXP(_ amount: Int) -> [Int] {
        var levelsGained: [Int] = []
        currentXP += amount
        while currentXP >= xpToNextLevel && currentLevel < 50 {
            currentXP -= xpToNextLevel
            currentLevel += 1
            levelsGained.append(currentLevel)
            checkAndUnlockFeatures()
        }
        if currentLevel == 50 { currentXP = 0 }
        return levelsGained
    }

    // MARK: Feature Unlocking

    func isFeatureUnlocked(_ feature: GameFeature) -> Bool {
        unlockedFeatures.contains(feature.rawValue)
    }

    mutating func checkAndUnlockFeatures() {
        for feature in GameFeature.allCases {
            guard !isFeatureUnlocked(feature) else { continue }
            let levelOK = currentLevel >= feature.requiredLevel
            let storyOK: Bool
            if let requiredAct = feature.requiredStoryAct {
                storyOK = storyAct >= requiredAct
            } else {
                storyOK = true
            }
            if levelOK && storyOK {
                unlockedFeatures.insert(feature.rawValue)
            }
        }
    }

    /// Called when a story act completes — re-evaluates story-gated features.
    mutating func advanceStoryAct(to act: Int) {
        storyAct = act
        checkAndUnlockFeatures()
    }
}

// MARK: - XP Curve

struct XPCurve {
    /// XP required to level up FROM the given level to level+1.
    static func xpRequired(forLevel level: Int) -> Int {
        switch level {
        case 1:       return 100
        case 2:       return 150
        case 3:       return 200
        case 4:       return 250
        case 5:       return 350
        case 6...9:   return 100 * level + 150
        default:      return 200 * level   // Linear from level 10+
        }
    }
}

// MARK: - XP Sources

enum XPSource {
    case feed, play, clean, sell, fulfillOrder, levelMilestone, breedSuccess

    var xpReward: Int {
        switch self {
        case .feed:          return 5
        case .play:          return 8
        case .clean:         return 5
        case .sell:          return 20
        case .fulfillOrder:  return 35
        case .levelMilestone: return 0
        case .breedSuccess:  return 50
        }
    }
}

File: KeeperLegacyGodot/Models/Progression.cs
// Models/Progression.cs
// Port of Progression.swift
// Pure C# — no Godot dependency.

using System;
using System.Collections.Generic;
using System.Linq;

namespace KeeperLegacy.Models
{
    // ── Feature Flags ─────────────────────────────────────────────────────────

    public enum GameFeature
    {
        Shop,
        Habitat,
        Feeding,
        Playing,
        Selling,
        CustomerOrders,
        Breeding,
        Monsterpedia,
        HabitatExpansion,
        Mutations,
        MagicalHabitat,
        Cosmetics
    }

    public static class GameFeatureExtensions
    {
        public static string RawValue(this GameFeature feature) => feature switch
        {
            GameFeature.Shop             => "shop",
            GameFeature.Habitat          => "habitat",
            GameFeature.Feeding          => "feeding",
            GameFeature.Playing          => "playing",
            GameFeature.Selling          => "selling",
            GameFeature.CustomerOrders   => "customerOrders",
            GameFeature.Breeding         => "breeding",
            GameFeature.Monsterpedia     => "monsterpedia",
            GameFeature.HabitatExpansion => "habitatExpansion",
            GameFeature.Mutations        => "mutations",
            GameFeature.MagicalHabitat   => "magicalHabitat",
            GameFeature.Cosmetics        => "cosmetics",
            _                            => feature.ToString().ToLower()
        };

        /// Minimum level required to unlock (0 = available from start).
        public static int RequiredLevel(this GameFeature feature) => feature switch
        {
            GameFeature.Shop             => 0,
            GameFeature.Habitat          => 0,
            GameFeature.Feeding          => 0,
            GameFeature.Playing          => 0,
            GameFeature.Selling          => 1,
            GameFeature.CustomerOrders   => 3,
            GameFeature.Breeding         => 12,
            GameFeature.Monsterpedia     => 2,
            GameFeature.HabitatExpansion => 2,
            GameFeature.Mutations        => 15,
            GameFeature.MagicalHabitat   => 0,   // Story-gated, not level-gated
            GameFeature.Cosmetics        => 5,
            _                            => 0
        };

        /// Story act required in addition to level (null = no story gate).
        public static int? RequiredStoryAct(this GameFeature feature) => feature switch
        {
            GameFeature.Breeding       => 1,
            GameFeature.MagicalHabitat => 2,
            _                          => null
        };

        public static GameFeature? FromRawValue(string raw) => raw switch
        {
            "shop"             => GameFeature.Shop,
            "habitat"          => GameFeature.Habitat,
            "feeding"          => GameFeature.Feeding,
            "playing"          => GameFeature.Playing,
            "selling"          => GameFeature.Selling,
            "customerOrders"   => GameFeature.CustomerOrders,
            "breeding"         => GameFeature.Breeding,
            "monsterpedia"     => GameFeature.Monsterpedia,
            "habitatExpansion" => GameFeature.HabitatExpansion,
            "mutations"        => GameFeature.Mutations,
            "magicalHabitat"   => GameFeature.MagicalHabitat,
            "cosmetics"        => GameFeature.Cosmetics,
            _                  => null
        };
    }

    // ── Milestone Rewards ─────────────────────────────────────────────────────

    public class MilestoneReward
    {
        public int Level { get; }
        public int CoinBonus { get; }
        public int StardustBonus { get; }
        public GameFeature? UnlocksFeature { get; }
        public string? StoryEvent { get; }
        public string Description { get; }

        public MilestoneReward(int level, int coinBonus, int stardustBonus,
            GameFeature? unlocksFeature, string? storyEvent, string description)
        {
            Level          = level;
            CoinBonus      = coinBonus;
            StardustBonus  = stardustBonus;
            UnlocksFeature = unlocksFeature;
            StoryEvent     = storyEvent;
            Description    = description;
        }

        public static readonly Dictionary<int, MilestoneReward> Milestones = new()
        {
            [5]  = new(5,  200,  0,   GameFeature.Cosmetics, null,                 "Cosmetics unlocked!"),
            [10] = new(10, 500,  25,  null,                  "act1_shop_secret",   "The shop's secret begins to reveal itself..."),
            [12] = new(12, 300,  0,   GameFeature.Breeding,  null,                 "Breeding unlocked! Requires Act I completion."),
            [15] = new(15, 400,  0,   GameFeature.Mutations, null,                 "Mutation variants discovered!"),
            [25] = new(25, 1000, 50,  null,                  "act2_revelation",    "A revelation about the ancient civilization..."),
            [50] = new(50, 5000, 200, null,                  "act3_legacy",        "You have mastered the Keeper's Legacy!"),
        };
    }

    // ── Player Progression State ──────────────────────────────────────────────

    public class PlayerProgression
    {
        public int CurrentLevel { get; set; }
        public int CurrentXP { get; set; }
        public HashSet<string> UnlockedFeatures { get; set; }   // GameFeature.RawValue()
        public HashSet<int> ClaimedMilestones { get; set; }     // Level numbers claimed
        public int StoryAct { get; set; }                       // 1, 2, or 3

        public PlayerProgression()
        {
            CurrentLevel     = 1;
            CurrentXP        = 0;
            StoryAct         = 1;
            ClaimedMilestones = new HashSet<int>();

            // Unlock all features that require level 0 and no story gate
            UnlockedFeatures = new HashSet<string>(
                Enum.GetValues<GameFeature>()
                    .Where(f => f.RequiredLevel() == 0 && f.RequiredStoryAct() == null)
                    .Select(f => f.RawValue())
            );
        }

        // For deserialization
        public PlayerProgression(int currentLevel, int currentXP,
            HashSet<string> unlockedFeatures, HashSet<int> claimedMilestones, int storyAct)
        {
            CurrentLevel      = currentLevel;
            CurrentXP         = currentXP;
            UnlockedFeatures  = unlockedFeatures;
            ClaimedMilestones = claimedMilestones;
            StoryAct          = storyAct;
        }

        // ── XP / Leveling ─────────────────────────────────────────────────────

        /// XP needed to advance from current level to the next.
        public int XpToNextLevel => XPCurve.XpRequired(CurrentLevel);

        /// Add XP and handle level-ups. Returns list of levels gained (usually 0 or 1).
        public List<int> AddXP(int amount)
        {
            var levelsGained = new List<int>();
            CurrentXP += amount;

            while (CurrentXP >= XpToNextLevel && CurrentLevel < 50)
            {
                CurrentXP    -= XpToNextLevel;
                CurrentLevel += 1;
                levelsGained.Add(CurrentLevel);
                CheckAndUnlockFeatures();
            }

            if (CurrentLevel == 50) CurrentXP = 0;
            return levelsGained;
        }

        // ── Feature Unlocking ─────────────────────────────────────────────────

        public bool IsFeatureUnlocked(GameFeature feature) =>
            UnlockedFeatures.Contains(feature.RawValue());

        public void CheckAndUnlockFeatures()
        {
            foreach (var feature in Enum.GetValues<GameFeature>())
            {
                if (IsFeatureUnlocked(feature)) continue;

                bool levelOk = CurrentLevel >= feature.RequiredLevel();
                bool storyOk = feature.RequiredStoryAct() is int reqAct
                    ? StoryAct >= reqAct
                    : true;

                if (levelOk && storyOk)
                    UnlockedFeatures.Add(feature.RawValue());
            }
        }

        /// Called when a story act completes — re-evaluates story-gated features.
        public void AdvanceStoryAct(int act)
        {
            StoryAct = act;
            CheckAndUnlockFeatures();
        }
    }

    // ── XP Curve ──────────────────────────────────────────────────────────────

    public static class XPCurve
    {
        /// XP required to level up FROM the given level to level+1.
        public static int XpRequired(int level) => level switch
        {
            1     => 100,
            2     => 150,
            3     => 200,
            4     => 250,
            5     => 350,
            >= 6 and <= 9 => 100 * level + 150,
            _     => 200 * level   // Linear from level 10+
        };
    }

    // ── XP Sources ────────────────────────────────────────────────────────────

    public enum XPSource
    {
        Feed,
        Play,
        Clean,
        Sell,
        FulfillOrder,
        LevelMilestone,
        BreedSuccess
    }

    public static class XPSourceExtensions
    {
        public static int XpReward(this XPSource source) => source switch
        {
            XPSource.Feed           => 5,
            XPSource.Play           => 8,
            XPSource.Clean          => 5,
            XPSource.Sell           => 20,
            XPSource.FulfillOrder   => 35,
            XPSource.LevelMilestone => 0,
            XPSource.BreedSuccess   => 50,
            _                       => 0
        };
    }
}
